### DISCUSSION TITLE
Chart updating using keyboard input (Trading strategy "debug" mode)

### DISCUSSION THREAD

**Discussion by KuSi833** [highfestiva/finplot#197] [Q&A] [answered]
First of all I want to express how glad I am to find this repo. I was surprised how few options for chatting there were online.
Their performance was either bad for larger charts, or they were too static, or simply looked bad.

Now for my issue:
I want to create a trading strategy backtester + debugger in the sense that I can step through price history candle by candle and look at how my algo would react at each timestamp instead of looking at it all at once.
I want it to work is such a way that when the user presses the left or right arrow the chart either goes one candle forwards or backwards.
My problem arises when it comes to updating the chart with newer data, because when fplt.show() is reached I can no longer pass it any input.
I could sort of get an implementation working with the timer__callback function, but it doesn't work very well.
What would be the best way to tackle this issue?

My current plan is to get some multithreading going where one thread keeps track of input and inserts it into a pipe, meanwhile another thread has the timer_callback running in which the update function checks whether there is anything in the pipe.

Does any better solution come to mind or is this the way to go?

Kind regards.

**Comment by highfestiva** [comment, marked as answer]
Perhaps overriding `fplt._key_pressed` is a nice way to start? Perhaps implemented something like this:
```python
    if ev.key() == QtCore.Qt.Key_Left:
        vb.pan_x(steps=-1)
    elif ev.key() == QtCore.Qt.Key_Right:
        vb.pan_x(steps=+1)
```
Lemmeknow if there are some bugs, don't think I ever tested the `steps` implementation (only used the `percent` parameter).

**Comment by KuSi833** [comment]
I'm looking for something more like updating the data on each key_press. 
The following is a proof of concept but I'll need to find a less intrusive way to do it. 
The way it works now is I give each CandlestickItem a further last_index property and the entire dataframe so it can be referenced from the KeyPress override method.
```python
elif ev.text() == 'a':
  candlestickChart = vb.win.axs[0].items[4]
  candlestickChart.last_index -= 1
  df_insert = candlestickChart.df.loc[0:candlestickChart.last_index, ['date', 'open', 'close', 'high', 'low']]
  candlestickChart.update_data(df_insert)
```

**Reply by highfestiva** [reply]
Store the data yourself and use that, don't use the chart's data.

**Reply by KuSi833** [reply]
I am storing the data but have difficulty giving the function called on key_press reference to that data because it creates a circular dependency. To avoid this I let every OHLC chart have its own data variable which it then indexes.

**Comment by KuSi833** [comment]
I didn't realize it was possible to use the financial chart as a widget inside of a larger PyQt application using the following:
```python
self.fplt_widget = pg.PlotWidget(plotItem=fplt.create_plot_widget(self.window()))
```

Now I can simply update the chart by calling `update_data(new_dataframe)` whenever some button is pressed, or as @highfestiva mentioned when a _key_press is pressed (if overridden).

This all seems kind of obvious now, but hopefully it helps someone along the way.
